cmd/ocp-build-data-enforcer/main.go

```diff
@@ -0,0 +1,252 @@
+package main
+
+import (
+	"bytes"
+	"errors"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+
+	"github.com/openshift/imagebuilder"
+	dockercmd "github.com/openshift/imagebuilder/dockerfile/command"
+	"github.com/pmezard/go-difflib/difflib"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sync/errgroup"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"sigs.k8s.io/yaml"
+
+	"github.com/openshift/ci-tools/pkg/github"
+)
+
+type options struct {
+	ocpBuildDataRepoDir string
+}
+
+func gatherOptions() *options {
+	o := &options{}
+	flag.StringVar(&o.ocpBuildDataRepoDir, "ocp-build-data-repo-dir", "../ocp-build-data", "The directory in which the ocp-build-data reposity is")
+	flag.Parse()
+	return o
+}
+func main() {
+	opts := gatherOptions()
+
+	configs, err := gatherAllOCPImageConfigs(opts.ocpBuildDataRepoDir)
+	if err != nil {
+		logrus.WithError(err).Fatal("Failed to gather all ocp image configs")
+	}
+
+	streamMap, err := readStreamMap(opts.ocpBuildDataRepoDir)
+	if err != nil {
+		logrus.WithError(err).Fatal("Failed to read streamMap")
+	}
+
+	for cfgIdx := range configs {
+		dereferenceStreams(&configs[cfgIdx], streamMap)
+	}
+
+	errGroup := &errgroup.Group{}
+	var counter int
+	for _, config := range configs {
+		counter++
+		config := config
+		errGroup.Go(func() error {
+			processDockerfile(config)
+			return nil
+		})
+	}
+	if err := errGroup.Wait(); err != nil {
+		logrus.WithError(err).Fatal("Processing failed")
+	}
+	logrus.Infof("Processed %d configs", counter)
+}
+
+func dereferenceStreams(config *ocpImageConfig, streamMap streamMap) {
+	if replacement, hasReplacement := streamMap[config.From.Stream]; hasReplacement && replacement.UpstreamImage != nil {
+		config.From.Stream = *replacement.UpstreamImage
+	}
+	for blder := range config.From.Builder {
+		if replacement, hasReplacement := streamMap[config.From.Builder[blder].Stream]; hasReplacement && replacement.UpstreamImage != nil {
+			config.From.Builder[blder].Stream = *replacement.UpstreamImage
+		}
+	}
+}
+
+func processDockerfile(config ocpImageConfig) {
+	orgRepo := config.orgRepo()
+	log := logrus.WithField("file", config.SourceFileName).WithField("org/repo", orgRepo)
+	split := strings.Split(orgRepo, "/")
+	if n := len(split); n != 2 {
+		log.Errorf("splitting orgRepo didn't yield 2 but %d results", n)
+		return
+	}
+	if split[0] == "openshift-priv" {
+		log.Warn("Ignoring repo in openshift-priv org")
```

Comment:
Not really a warning, maybe not even worth logging

Reply:
Also, for e.g. `openshift/kubernetes` isn't the only config in `ocp-build-data` for the private fork?

Reply:
I don't understand what you mean by the second comment, thats irrelevant for the tool or not? It will not process repos that are not referenced in the ocp-build-data repo

Reply:
I mean we cannot skip private things if they are configured in ocp-build-data and we *might* need to map private --> public, too, if it's only the private fork that is configured

Reply:
In this case, we would need DPTP to act on the public fork & branch. The translation is done by seeing if the ocp-build-data URL has a `public_upstreams` [mapping](https://github.com/openshift/ocp-build-data/blob/137f732fb8832043df20eac2063ffc5924220545/group.yml#L80-L85) which applies. The logic for translating private -> public is the longest prefix matches. The implementation in python can be found here: https://github.com/openshift/doozer/blob/e00d5fb5de8c934f6b6db9ab920dce3a86631901/doozerlib/runtime.py#L901-L942